### PR TITLE
Flink: Add FlinkWriterFactory

### PR DESCRIPTION
**Pull request by aokolnychyi**
This PR adds `FlinkWriterFactory` that extends `BaseWriterFactory` introduced in #2873.

### CODE REVIEW

**Comment by aokolnychyi** [comment]
cc @stevenzwu @openinx

**Comment by aokolnychyi** [comment]
I need a writer factory for Flink so that I can introduce new task writers.

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java [hunk 1]
```diff
@@ -0,0 +1,233 @@
...
+    builder.createWriterFunc(msgType -> FlinkParquetWriters.buildWriter(positionDeleteFlinkType(), msgType));
+    builder.transformPaths(path -> StringData.fromString(path.toString()));
+  }
+
+  @Override
+  protected void configureDataWrite(ORC.DataWriteBuilder builder) {
+    builder.createWriterFunc((iSchema, typDesc) -> FlinkOrcWriter.buildWriter(dataFlinkType(), iSchema));
+  }
+
+  private RowType dataFlinkType() {
```
Comment (stevenzwu) [inline flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:112]: should we move this logic to constructor?

Comment (stevenzwu) [reply flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:112]: or maybe move this part to the Builder#build() method

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java [hunk 2]
```diff
@@ -0,0 +1,233 @@
...
+      this.dataFileFormat = newDataFileFormat;
+      return this;
+    }
+
+    Builder dataSchema(Schema newDataSchema) {
+      this.dataSchema = newDataSchema;
+      return this;
+    }
+
+    Builder dataFlinkType(RowType newDataFlinkType) {
```
Comment (stevenzwu) [inline flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:176]: Is this an optional param? In the other comment, I saw we are converting `Schema` (above) to `RowType` if `dataFlinkType` is not set.

I assume other params are all required. if so, maybe add a Javadoc to mark this as optional.

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java [hunk 3]
```diff
@@ -0,0 +1,233 @@
...
+    if (equalityDeleteFlinkType == null) {
+      Preconditions.checkNotNull(equalityDeleteRowSchema(), "Equality delete schema must not be null");
+      this.equalityDeleteFlinkType = FlinkSchemaUtil.convert(equalityDeleteRowSchema());
+    }
+
+    return equalityDeleteFlinkType;
+  }
+
+  private RowType positionDeleteFlinkType() {
+    if (positionDeleteFlinkType == null) {
```
Comment (stevenzwu) [inline flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:131]: similar comment as above

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java (hunk 1 above)
Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:112]: These are optional and may not be set/needed unless we actually write data files.
Pure DELETEs will not need it, for example.

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java (hunk 3 above)
Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:131]: This one is similar to data files. The value is optional and may not be needed.

File: flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java (hunk 2 above)
Comment (aokolnychyi) [reply flink/src/main/java/org/apache/iceberg/flink/sink/FlinkWriterFactory.java:181]: Added docs to optional methods.

**Review by stevenzwu — APPROVED** [review]
nice base test class

**Comment by aokolnychyi** [comment]
Thanks for reviewing, @openinx @stevenzwu!

**Pull request merged**